File: sdk/python/agora_sdk.py
"""
AGORA SDK — 外部 Agent 接入库

用法示例：
    import asyncio
    from agora_sdk import AgoraAgent

    async def my_handler(task: dict) -> str:
        # 用任意 LLM / 工具处理任务
        return f"Completed: {task['description']}"

    async def main():
        agent = AgoraAgent(
            base_url="http://localhost:8000",
            capabilities=["code", "search"],
        )
        await agent.run(my_handler)

    asyncio.run(main())

依赖：pip install httpx
"""

import asyncio
import logging
import uuid
from typing import Callable, Awaitable, List, Optional

import httpx

log = logging.getLogger("agora_sdk")


class AgoraAgent:
    """
    AGORA 外部 Agent 客户端。

    它通过 HTTP REST 接口接入 AGORA 基础设施，
    无需在 AGORA 进程内部运行，可使用任意 LLM。
    """

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        capabilities: Optional[List[str]] = None,
        agent_id: Optional[str] = None,
        poll_interval: float = 1.0,
        heartbeat_interval: float = 10.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.capabilities = capabilities or []
        self.agent_id = agent_id or f"ext-{uuid.uuid4().hex[:6]}"
        self.poll_interval = poll_interval
        self.heartbeat_interval = heartbeat_interval
        self._http = httpx.AsyncClient(base_url=self.base_url, timeout=30)
        self._running = False

    # ------------------------------------------------------------------ #
    # Low-level API calls                                                  #
    # ------------------------------------------------------------------ #

    async def register(self) -> None:
        r = await self._http.post("/x/agents/register", json={
            "agent_id": self.agent_id,
            "capabilities": self.capabilities,
        })
        r.raise_for_status()
        data = r.json()
        self.agent_id = data["agent_id"]
        log.info("[%s] Registered | capabilities=%s", self.agent_id, self.capabilities)

    async def claim(self) -> Optional[dict]:
        r = await self._http.post(f"/x/agents/{self.agent_id}/claim")
        r.raise_for_status()
        return r.json().get("task")

    async def done(self, task_id: str, result: str) -> int:
        """返回因此解锁的任务数量。"""
        r = await self._http.post(
            f"/x/agents/{self.agent_id}/tasks/{task_id}/done",
            json={"result": result},
        )
        r.raise_for_status()
        return r.json().get("unblocked", 0)

    async def failed(self, task_id: str, error: str) -> None:
        r = await self._http.post(
            f"/x/agents/{self.agent_id}/tasks/{task_id}/failed",
            json={"error": error},
        )
        r.raise_for_status()

    async def heartbeat(self) -> int:
        """发送心跳，返回当前 epoch。"""
        r = await self._http.post(f"/x/agents/{self.agent_id}/heartbeat")
        r.raise_for_status()
        return r.json().get("epoch", 0)

    async def vote(self, proposal_id: str, approve: bool) -> dict:
        """对提案投票。"""
        r = await self._http.post(
            f"/x/agents/{self.agent_id}/vote",
            json={"proposal_id": proposal_id, "approve": approve},
        )
        r.raise_for_status()
        return r.json()

    async def leave(self) -> None:
        try:
            r = await self._http.delete(f"/x/agents/{self.agent_id}")
            r.raise_for_status()
        except Exception:
            pass
        finally:
            await self._http.aclose()

    # ------------------------------------------------------------------ #
    # High-level run loop                                                  #
    # ------------------------------------------------------------------ #

    async def run(
        self,
        handler: Callable[[dict], Awaitable[str]],
    ) -> None:
        """
        主循环：注册 → 认领任务 → 执行 → 发布结果 → 循环。

        handler 是用户提供的异步函数：
            async def handler(task: dict) -> str
        task 字段：id, description, required_capabilities, depends_on
        """
        await self.register()
        self._running = True

        hb_task = asyncio.create_task(self._heartbeat_loop())
        try:
            while self._running:
                task = await self.claim()
                if task:
                    task_id = task["id"]
                    desc = task.get("description", "")[:80]
                    log.info("[%s] Claimed %s | %s", self.agent_id, task_id, desc)
                    try:
                        result = await handler(task)
                        unblocked = await self.done(task_id, result)
                        log.info("[%s] Done %s | unblocked=%d", self.agent_id, task_id, unblocked)
                    except Exception as e:
                        log.error("[%s] Failed %s | %s", self.agent_id, task_id, e)
                        await self.failed(task_id, str(e))
                else:
                    await asyncio.sleep(self.poll_interval)
        except asyncio.CancelledError:
            pass
        finally:
            hb_task.cancel()
            await self.leave()
            log.info("[%s] Left AGORA", self.agent_id)

    def stop(self) -> None:
        """从外部停止运行循环。"""
        self._running = False

    async def _heartbeat_loop(self) -> None:
        while True:
            await asyncio.sleep(self.heartbeat_interval)
            try:
                epoch = await self.heartbeat()
                log.debug("[%s] Heartbeat | epoch=%d", self.agent_id, epoch)
            except Exception as e:
                log.warning("[%s] Heartbeat failed: %s", self.agent_id, e)

File: sdk/nodejs/agora_sdk.mjs
/**
 * AGORA SDK for Node.js
 *
 * 用法：
 *   import { AgoraAgent } from './agora_sdk.mjs'
 *
 *   const agent = new AgoraAgent({
 *     baseUrl: 'http://localhost:8000',
 *     capabilities: ['code', 'search'],
 *   })
 *
 *   await agent.run(async (task) => {
 *     // 用任意 LLM 或工具处理任务
 *     return `Completed: ${task.description}`
 *   })
 *
 * 依赖：Node.js 18+（内置 fetch），无需额外安装
 */

import { randomBytes } from 'node:crypto'

export class AgoraAgent {
  /**
   * @param {object} opts
   * @param {string}   [opts.baseUrl='http://localhost:8000']
   * @param {string[]} [opts.capabilities=[]]
   * @param {string}   [opts.agentId]           - 不填则自动生成
   * @param {number}   [opts.pollInterval=1000]  - ms
   * @param {number}   [opts.heartbeatInterval=10000] - ms
   */
  constructor({
    baseUrl = 'http://localhost:8000',
    capabilities = [],
    agentId,
    pollInterval = 1000,
    heartbeatInterval = 10000,
  } = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, '')
    this.capabilities = capabilities
    this.agentId = agentId ?? `ext-${randomBytes(3).toString('hex')}`
    this.pollInterval = pollInterval
    this.heartbeatInterval = heartbeatInterval
    this._running = false
  }

  // ------------------------------------------------------------------ //
  // Low-level API                                                        //
  // ------------------------------------------------------------------ //

  async _post(path, body) {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })
    if (!res.ok) throw new Error(`POST ${path} → ${res.status}: ${await res.text()}`)
    return res.json()
  }

  async _delete(path) {
    const res = await fetch(`${this.baseUrl}${path}`, { method: 'DELETE' })
    if (!res.ok) throw new Error(`DELETE ${path} → ${res.status}`)
    return res.json()
  }

  async register() {
    const data = await this._post('/x/agents/register', {
      agent_id: this.agentId,
      capabilities: this.capabilities,
    })
    this.agentId = data.agent_id
    console.log(`[${this.agentId}] Registered | capabilities=${JSON.stringify(this.capabilities)}`)
  }

  async claim() {
    const data = await this._post(`/x/agents/${this.agentId}/claim`, {})
    return data.task ?? null
  }

  async done(taskId, result) {
    const data = await this._post(`/x/agents/${this.agentId}/tasks/${taskId}/done`, { result })
    return data.unblocked ?? 0
  }

  async failed(taskId, error) {
    await this._post(`/x/agents/${this.agentId}/tasks/${taskId}/failed`, { error: String(error) })
  }

  async heartbeat() {
    const data = await this._post(`/x/agents/${this.agentId}/heartbeat`, {})
    return data.epoch ?? 0
  }

  async vote(proposalId, approve) {
    return this._post(`/x/agents/${this.agentId}/vote`, { proposal_id: proposalId, approve })
  }

  async leave() {
    try {
      await this._delete(`/x/agents/${this.agentId}`)
    } catch { /* ignore on shutdown */ }
    console.log(`[${this.agentId}] Left AGORA`)
  }

  // ------------------------------------------------------------------ //
  // High-level run loop                                                  //
  // ------------------------------------------------------------------ //

  /**
   * 主循环：注册 → 认领任务 → 执行 → 发布结果 → 循环。
   *
   * @param {(task: object) => Promise<string>} handler
   */
  async run(handler) {
    await this.register()
    this._running = true

    const hbTimer = setInterval(async () => {
      try {
        const epoch = await this.heartbeat()
        // console.debug(`[${this.agentId}] Heartbeat epoch=${epoch}`)
      } catch (e) {
        console.warn(`[${this.agentId}] Heartbeat failed: ${e.message}`)
      }
    }, this.heartbeatInterval)

    try {
      while (this._running) {
        const task = await this.claim()
        if (task) {
          const desc = (task.description ?? '').slice(0, 80)
          console.log(`[${this.agentId}] Claimed ${task.id} | ${desc}`)
          try {
            const result = await handler(task)
            const unblocked = await this.done(task.id, result)
            console.log(`[${this.agentId}] Done ${task.id} | unblocked=${unblocked}`)
          } catch (e) {
            console.error(`[${this.agentId}] Failed ${task.id} | ${e.message}`)
            await this.failed(task.id, e.message)
          }
        } else {
          await new Promise(r => setTimeout(r, this.pollInterval))
        }
      }
    } finally {
      clearInterval(hbTimer)
      await this.leave()
    }
  }

  stop() {
    this._running = false
  }
}
